### PR TITLE
Implement gradient for vector repetitions

### PR DESCRIPTION
**Pull request by ricardoV94**
Nobody asked for it, and nobody probably ever needed it (it exists since 2012), but it was a fun challenge.

Also cleaned up a bit the implementation by not allowing negative axis on the Op itself like we do in other cases (the helper repeat handles the conversion to positive axis)

doctest fails due to an unrelated print that should be solved by #1193 

----
📚 Documentation preview 📚: https://pytensor--1192.org.readthedocs.build/en/1192/

### CODE REVIEW

File: pytensor/tensor/extra_ops.py [hunk 1]
```diff
@@ -687,58 +692,64 @@ def make_node(self, x, repeats):
...
-            ]
         elif repeats.ndim == 1:
-            # For this implementation, we would need to specify the length
-            # of repeats in order to split gz in the right way to sum
-            # the good part.
-            raise NotImplementedError()
+            # To sum the gradients that belong to the same repeated x,
+            # We create a repeated eye and dot product it with the gradient.
+            axis_size = x.size if axis is None else x.shape[axis]
+            tiled_eye = repeat(
```
Comment (jessegrabowski) [inline pytensor/tensor/extra_ops.py:722]: This can use `_iota` from the einsum code I think?

Also why is it called `tiled_eye` but it's repeated? `tiled_eye` should use `tile` : )

File: pytensor/tensor/extra_ops.py [hunk 2]
```diff
@@ -687,58 +692,64 @@ def make_node(self, x, repeats):
...
+            )  # A sparse repeat would be neat
+
+            if axis is None:
+                gx = gz @ tiled_eye
+                # Undo the ravelling when axis=None
+                gx = gx.reshape(x.shape)
+            else:
+                # Place gradient axis at end for dot product
+                gx = ptb.moveaxis(gz, axis, -1)
+                gx = gx @ tiled_eye
```
Comment (jessegrabowski) [inline pytensor/tensor/extra_ops.py:733]: We have a tensor_dot or dot_general or something that handles these kinds of move + dot? You could probably just einsum here as well.

File: pytensor/tensor/extra_ops.py [hunk 3]
```diff
@@ -751,82 +762,97 @@ def infer_shape(self, fgraph, node, ins_shapes):
...
+                out_shape[axis] = out_shape[axis] * repeats
             else:
-                out_shape[self.axis] = pt_sum(repeats, dtype=dtype)
+                out_shape[axis] = pt_sum(repeats, dtype=dtype)
         return [out_shape]
 
 
-def repeat(x, repeats, axis=None):
-    """Repeat elements of an array.
+def repeat(a: "TensorLike", repeats: TensorLike, axis: int or None) -> TensorVariable:
```
Comment (jessegrabowski) [inline pytensor/tensor/extra_ops.py:771]: Why is one typehint a string but the other is not?

File: pytensor/tensor/extra_ops.py [hunk 4]
```diff
@@ -751,82 +762,97 @@ def infer_shape(self, fgraph, node, ins_shapes):
...
+    """Repeat elements of a tensor.
 
-    It returns an array which has the same shape as `x`, except along the given
-    `axis`. The `axis` parameter is used to specify the axis along which values
-    are repeated. By default, a flattened version of `x` is used.
+    See `numpy.repeat` for more information.
```
Comment (jessegrabowski) [inline pytensor/tensor/extra_ops.py:774]: What if I'm working in pure pytensor? I have to import numpy just to read our docs?

File: pytensor/tensor/extra_ops.py [hunk 5]
```diff
@@ -751,82 +762,97 @@ def infer_shape(self, fgraph, node, ins_shapes):
...
-    The number of repetitions for each element is `repeats`.  `repeats` is
-    broadcasted to fit the length of the given `axis`.
 
     Parameters
     ----------
-    x
-        Input data, tensor variable.
-    repeats
-        int, scalar or tensor variable
+    a: tensor_like
```
Comment (jessegrabowski) [inline pytensor/tensor/extra_ops.py:779]:
```suggestion
    a: TensorLike
```
I think the pycharm linter checks these for valid types. There's no type called `tensor_like`, and it's also not fully human readable; kind of a weird middle ground

File: pytensor/tensor/extra_ops.py [hunk 6]
```diff
@@ -751,82 +762,97 @@ def infer_shape(self, fgraph, node, ins_shapes):
...
+    Returns
+    -------
+    repeated_tensor: TensorVariable
+        Output tensor which as the same shape as a, except along the given axis
+
+    Examples
     --------
-    tensor.tile
+
+    .. testcode::
```
Comment (jessegrabowski) [inline pytensor/tensor/extra_ops.py:794]: Add an example with `axis=None`, because the flattening behavior will surprise users. I remember getting got by it in `cumsum`.

File: pytensor/tensor/extra_ops.py [hunk 7]
```diff
@@ -751,82 +762,97 @@ def infer_shape(self, fgraph, node, ins_shapes):
...
     if repeats.ndim == 1 and not repeats.broadcastable[0]:
-        return Repeat(axis=axis)(x, repeats)
+        # We only use the Repeat Op for vector repeats
+        return Repeat(axis=axis)(a, repeats)
     else:
         if repeats.ndim == 1:
             repeats = repeats[0]
 
-        if x.dtype == "uint64":
+        if a.dtype == "uint64":
```
Comment (jessegrabowski) [inline pytensor/tensor/extra_ops.py:831]: Why this one specific dtype? `uint32` is OK?

**Review by jessegrabowski — CHANGES_REQUESTED** [review]
Needs some small cleanup, otherwise cool!

File: pytensor/tensor/extra_ops.py (hunk 4 above)
Comment (ricardoV94) [reply pytensor/tensor/extra_ops.py:774]: you go to the docs :)

Comment (ricardoV94) [reply pytensor/tensor/extra_ops.py:774]: We could add links like JAX does, I think we already have cross-ref so it should work the same way: https://github.com/jax-ml/jax/blob/e7acb20ea2ba037a0e82134a6b1bcf35430d28ca/jax/_src/numpy/lax_numpy.py#L6980

Then the docs have a clickable link: https://docs.jax.dev/en/latest/_autosummary/jax.numpy.repeat.html

File: pytensor/tensor/extra_ops.py (hunk 5 above)
Comment (ricardoV94) [reply pytensor/tensor/extra_ops.py:779]: I thought we had a glossary like in PyMC: https://www.pymc.io/projects/docs/en/stable/glossary.html#term-tensor_like

In which case we should use `tensor_like` in the docs

File: pytensor/tensor/extra_ops.py (hunk 1 above)
Comment (ricardoV94) [reply pytensor/tensor/extra_ops.py:722]: `_iota` is just an arange, no? I need a way to combine entries that belong to the same repeat, I don't think any from of advanced indexing gives me that

File: pytensor/tensor/extra_ops.py (hunk 2 above)
Comment (ricardoV94) [reply pytensor/tensor/extra_ops.py:733]: tensordot is not what I want as it mixes all the dimensions. dot_general allows me to user tensordot + vectorize to achieve the same but:
1. It ends up being more verbose because I have list all core and batch axes
2. It calls tensordot which for a simple matmul creates an unnecessarily complex graph full of reshapes. This is alleviated by #1123 (but not always). I'll open some issues to get shortcuts on tensordot.

Comment (jessegrabowski) [reply pytensor/tensor/extra_ops.py:733]: It just seems we should have tools for this kind of move + dot + move, and if our existing tools suck we should address that

Comment (ricardoV94) [reply pytensor/tensor/extra_ops.py:733]: https://github.com/pymc-devs/pytensor/issues/1201

File: pytensor/tensor/extra_ops.py (hunk 1 above)
Comment (ricardoV94) [reply pytensor/tensor/extra_ops.py:722]: Fixed the `tiled_eye` name

File: pytensor/tensor/extra_ops.py (hunk 3 above)
Comment (ricardoV94) [reply pytensor/tensor/extra_ops.py:771]: No idea, doesn't seem like it needs to be a string

File: pytensor/tensor/extra_ops.py (hunk 6 above)
Comment (ricardoV94) [reply pytensor/tensor/extra_ops.py:794]: Added

File: pytensor/tensor/extra_ops.py (hunk 7 above)
Comment (ricardoV94) [reply pytensor/tensor/extra_ops.py:831]: Yes, this limitation is for another reason, since we don't use repeat at all, but we use reshape with `repeat_shape[axis] * repeats` and funnily enough if you multiply an int64 (shape) and an uint64 (potential repeat), you get a float64, which is not a valid type for the reshape operation. I added a comment for that. We are testing all the valid/invalid cases in `TestRepeat.test_basic` in any case

File: pytensor/tensor/extra_ops.py (hunk 1 above)
Comment (ricardoV94) [reply pytensor/tensor/extra_ops.py]: Right, but I don't need arbitrary identity matrices, 2d is fine because repeats is just a vector.

File: pytensor/tensor/extra_ops.py (hunk 3 above)
Comment (ricardoV94) [reply pytensor/tensor/extra_ops.py]: Fixed

File: pytensor/tensor/extra_ops.py (hunk 2 above)
Comment (ricardoV94) [reply pytensor/tensor/extra_ops.py]: I agree, a big part of it has to do with reshape, which is an awful Op. Issues #201 and discussion #883 are directly on that.

**Review by jessegrabowski — APPROVED** [review]
